src/Core/Geometry/Volume.hpp

```diff
@@ -0,0 +1,300 @@
+
+    /// Get the value of the given bin
+    inline Utils::optional<ValueType> getBinValue( Eigen::Ref<const IndexType> p ) const {
+        if ( auto res = linearIndex( p ) ) return getBinValue( *res );
+        return {};
+    }
+
+    /** Get the function value at a given position p (discrete implementation).
+     * \warning no bounds checking on the parameter p
+     * \see #getBinValue
+     */
+    inline Utils::optional<ValueType> getValue( Eigen::Ref<const Vector3> p ) const override final {
+        return getBinValue( ( p.cwiseQuotient( m_binSize ) ).cast<typename IndexType::Scalar>() );
+    }
+
+    /// Increment bin p by value
+    /// \note : do nothing if p is out of bounds
+    inline bool addToBin( const ValueType& value, Eigen::Ref<const IndexType> p ) {
+        if ( auto res = linearIndex( p ) )
+        {
+            addToBin( value, *res );
+            return true;
+        }
+        return false;
+    }
+
+  protected:
+    /// Convert the 3D position into a linear index on the bin set
+    inline Utils::optional<typename IndexType::Scalar>
+    linearIndex( Eigen::Ref<const IndexType> p ) const {
+        using Integer = typename IndexType::Scalar;
+        if ( ( p.array() >= m_size.array() ).any() ) return {};
+        return p.dot( IndexType( Integer( 1 ), m_size( 0 ), m_size( 0 ) * m_size( 1 ) ) );
+    }
+    /// Get the bin value
+    virtual Utils::optional<ValueType> getBinValue( typename IndexType::Scalar idx ) const = 0;
+    /// Add a vlue to the bin
+    virtual void addToBin( const ValueType& value, typename IndexType::Scalar idx )        = 0;
+
+    /// Method called when size as been updated
+    virtual void updateStorage() = 0;
+
+  private:
+    IndexType m_size;  /// <\brief number of bins per dimension
+    Vector3 m_binSize; /// <\brief size of a bin per dimension
+};
+
+/// Discrete volume data storing values in a regular grid
+class RA_CORE_API VolumeGrid : public AbstractDiscreteVolume
+{
+  public:
+    using ValueType = AbstractDiscreteVolume::ValueType;
+    using IndexType = AbstractDiscreteVolume::IndexType;
+
+    using Container = std::vector<ValueType>;
+
+  public:
+    inline VolumeGrid( const ValueType& defaultValue = ValueType( 0. ) ) :
+        AbstractDiscreteVolume( DISCRETE_DENSE ), m_defaultValue( defaultValue ) {}
+    VolumeGrid( const VolumeGrid& data ) = default;
+    VolumeGrid& operator=( const VolumeGrid& ) = default;
+    ~VolumeGrid() override                     = default;
+
+    using AbstractDiscreteVolume::addToBin;
+    using AbstractDiscreteVolume::getBinValue;
+
+    /// Direct access to the managed data
+    inline const Container& data() const { return m_data; }
+    /// Direct access, with modification alllowed to the managed data
+    inline Container& data() { return m_data; }
+
+    /// Add a value to all bins
+    inline void addToAllBins( const ValueType& value ) {
+        for ( auto& v : m_data ) {
+            v += value;
```

Comment:
No :

Reply:
No. If we want to modernize the code, instead of relying on old writings, we can do this ```
std::transform( std::execution::par,
                m_data.begin(), m_data.end(), m_data.begin(),
                [value](auto v) { return v+value; });
```
Not better supported than OpenMP and range for but ready for the next.

Reply:
Well so openmp or not ?

Reply:
In my sense yes, since the stl parallel execution policies do not seem to be wildly supported yet

Reply:
let's decide that later, in a more general way about how to handle parallelism in radium.